lldpd: Init script update to support full range of supported LLDPD options

howels opened the pull request:
* LLDPD will now bind only to the specified interfaces.
* LLDPD now supports "LLDP Med Fast Start".
* LLDP location parameter is now fully implemented and respects the uci setting.
* LLDPD socket is now closed on exit.
* LLDPD now supports enabling or disabling CDP protocol as well as LLDP.
* LLDP location default is now changed to generic location. Please see LLDPD man page for details of these feaures.

Update of https://github.com/openwrt/openwrt/pull/13018

ptpt52 commented:
mark to follow

howels commented:
First PR here. I have read the patch guidelines and contributor guide but if I have missed something please let me know.

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -19,17 +26,85 @@ find_release_info()
+	local lldp_tx_interval
+	local lldp_tx_hold
+	local lldp_mgmt_ip
+	local lldp_agetntype
```
This PR carried over the same mistakes as @marek22k - see my branch for the relevant corrections:

https://github.com/openwrt/openwrt/compare/main...systemcrash:openwrt:lldp_fixes3

i.e. change to `lldap_agenttype`

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -130,5 +326,6 @@ reload_service() {
 
 stop_service() {
-	rm -rf /var/run/lldp $LLDPSOCKET
+	rm -rf ${LLDPD_RUN} $LLDPSOCKET
```
`rm -rf ${LLDPD_RUN} $LLDPSOCKET 2>/dev/null`

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -42,80 +117,201 @@ write_lldpd_conf()
+		procd_append_param command '-ll'
+	fi
+
+	if [ $enable_cdp -gt 0 ]; then
+		if [ $cdp_version == "cdpv1v2" ]; then
+			if [ $force_cdp -gt 0 ]; then
+				if [ $force_cdpv2 -gt 0 ]; then
+					# CDPv1 enabled, CDPv2 forced
+					procd_append_param command '-ccc'
+				else
+					# CDPv1 forced, CDPv2 enabled
+					procd_append_param command '-cc'
+				fi
+			else
+				# CDPv1 and CDPv2 enabled
+				procd_append_param command '-c'
+			fi
+		elif [ $cdp_version == "cdpv2" ]; then
+			if [ $force_cdp -gt 0 ]; then
+				# CDPv1 disabled, CDPv2 forced
+				procd_append_param command '-ccccc'
+			else
+				# CDPv1 disabled, CDPv2 enabled
+				procd_append_param command '-cccc'
+			fi
+		fi
+	fi
```
Perhaps this could be simplified to:
```sh
if [ $enable_cdp -gt 0 ]; then
	if [ $cdp_version == "cdpv1v2" ]; then
		if [ $force_cdp -gt 0 ] && [ $force_cdpv2 -gt 0 ]; then
			procd_append_param command '-ccc'
		elif [ $force_cdp -gt 0 ]; then
			procd_append_param command '-cc'
		else
			procd_append_param command '-c'
		fi
	elif [ $cdp_version == "cdpv2" ]; then
		if [ $force_cdp -gt 0 ]; then
			procd_append_param command '-ccccc'
		else
			procd_append_param command '-cccc'
		fi
	fi
fi

```

systemcrash replied:
Maybe even something more aesthetic:
```sh
if [ $enable_cdp -gt 0 ]; then
	if [ $cdp_version == "cdpv2" ]; then
		if [ $force_cdp -gt 0 ]; then
			procd_append_param command '-ccccc'
		else
			procd_append_param command '-cccc'
		fi
	elif [ $cdp_version == "cdpv1v2" ]; then
		if [ $force_cdp -gt 0 ] && [ $force_cdpv2 -gt 0 ]; then
			procd_append_param command '-ccc'
		elif [ $force_cdp -gt 0 ]; then
			procd_append_param command '-cc'
		else
			procd_append_param command '-c'
		fi
	fi
fi

```

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -42,80 +117,201 @@ write_lldpd_conf()
+			fi
+		fi
+	fi
+	
+	if [ $enable_fdp -gt 0 ]; then
+		if [ $force_fdp -gt 0 ]; then
+			procd_append_param command '-ff'
+		else
+			procd_append_param command '-f'
+		fi
+	fi
```
Likewise perhaps:
```sh
if [ $enable_fdp -gt 0 ] && [ $force_fdp -gt 0 ]; then
	procd_append_param command '-ff'
elif [ $enable_fdp -gt 0 ]; then
	procd_append_param command '-f'
fi
```

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -19,17 +26,85 @@ find_release_info()
+	config_get lldp_syscapabilities 'config' 'lldp_syscapabilities'
+	config_get_bool lldpmed_fast_start 'config' 'lldpmed_fast_start' 0
+	config_get lldpmed_fast_start_tx_interval 'config' 'lldpmed_fast_start_tx_interval' 0
+	config_get lldp_syscapabilities 'config' 'lldp_syscapabilities'
```
duplicate. remove. See my version...

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -19,17 +26,85 @@ find_release_info()
+	local lldp_mgmt_ip
+	local lldp_agetntype
+	local lldp_portidsubtype
+	local lldp_mgmt_ip
```
*sigh* duplicate. remove. See my version...

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -19,17 +26,85 @@ find_release_info()
+	config_get lldp_portidsubtype 'config' 'lldp_portidsubtype' 'macaddress'
+	config_get lldp_mgmt_ip 'config' 'lldp_mgmt_ip'
+	config_get lldp_syscapabilities 'config' 'lldp_syscapabilities'
+	config_get_bool lldpmed_fast_start 'config' 'lldpmed_fast_start' 0
```
Fuck. Why. Why `lldpmed_fast_start` but also `lldp_fast_start` which both do at least one of the same things?

systemcrash replied:
Must reconcile these two: retain the one that openwrt uses today, and integrate the tanto one. I think we must dop `lldp_fast_start`.

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -42,80 +117,201 @@ write_lldpd_conf()
+	[ -n "$lldp_platform" ] && echo "configure system platform" "\"$lldp_platform\"" >> "$LLDPD_CONF"
 	[ -n "$lldp_mgmt_ip" ] && echo "configure system ip management pattern" "\"$lldp_mgmt_ip\"" >> "$LLDPD_CONF"
 	[ -n "$lldp_syscapabilities" ] && echo "configure system capabilities enabled" "\"$lldp_syscapabilities\"" >> "$LLDPD_CONF"
+	[ $lldp_fast_start -gt 0 ] echo "configure med fast-start" "enable" >> "$LLDPD_CONF"
```
...

systemcrash commented:
Please also rename this PR to something more apt.

howels commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -19,17 +26,85 @@ find_release_info()
+	config_get lldp_agenttype 'config' 'lldp_agenttype' 'nearest-bridge'
+	config_get lldp_portidsubtype 'config' 'lldp_portidsubtype' 'macaddress'
+	config_get lldp_mgmt_ip 'config' 'lldp_mgmt_ip'
+	config_get lldp_syscapabilities 'config' 'lldp_syscapabilities'
```
@systemcrash I noticed that your branch doesn't configure `lldp_syscapabilities` - is that intentional?

systemcrash replied:
I don't know where you got that from: my branch leaves `lldp_syscapabilities` untouched. `lldp_syscapabilities` is configured. Line 48-49 + 57.

systemcrash replied:
My intention was to remove as much jumble from the noise of the initial PR, to retain focus on what actually (needed to be) changed. You're welcome to include it, but it's just moving lines around which takes a reviewer time to untangle.

howels replied:
I must have missed that, I'll recheck.

howels commented:
> Please also rename this PR to something more apt.

Hope the new title is better.

systemcrash commented:
OK, I think we can bring in some other reviewers. @hauke @stintel @PolynomialDivision

stintel commented:
Each of these changes should go in a **separate** commit. In its current form, it is impossible to review. Another reason: if any change causes a regression, unless someone comes up with a quick fix, the course of action is to git revert. This would then revert all changes. If each change is in a separate commit, the revert would be much smaller and all the other changes would still be there.

howels commented:
> Each of these changes should go in a **separate** commit. In its current form, it is impossible to review. Another reason: if any change causes a regression, unless someone comes up with a quick fix, the course of action is to git revert. This would then revert all changes. If each change is in a separate commit, the revert would be much smaller and all the other changes would still be there.

Sorry, I was trying to follow the OpenWRT guidelines. First rule on the Patch Checklist is "single commit (multiple commits must first be squashed)" https://openwrt.org/submitting-patches#patch_checklist

I can split this into multiple commits if that isn't the correct interpretation for patch submissions. However the scope is limited to addition of get/set options for a single config file so I was treating this as a single fix.

systemcrash commented:
> Each of these changes should go in a **separate** commit. In its current form, it is impossible to review. Another reason: if any change causes a regression, unless someone comes up with a quick fix, the course of action is to git revert. This would then revert all changes. If each change is in a separate commit, the revert would be much smaller and all the other changes would still be there.

Thanks for your thoughts @stintel

Impossible. I don't yet understand what part makes it impossible to review.

Lest we forget this is battle tested code in Tanto. I really don't see any advantage to split up the beef of what is meaningful to bring in for this changeset. They're all documented options from lldpd.

They've been tested against master.

Have you some lines drawn where one should split this up? Just each flag in a separate commit? ( And the init script respectively chopped up?)

I don't think many would have the patience to bisect to find which commit is the problem anyway, despite this being the principled way to go. So likely the whole PR would be reverted anyway, should there be any problem. Perhaps this might help the original committer @howels

howels commented:
I was hoping that as this is relatively simple shell script it would be readable as a single change. I can break each configuration option into a different commit but I would actually find a series of commits harder to review and the documented guidelines appeared to suggest that a patch (which I take to be analogous to a PR) should be a single commit.
If the ask is to submit multiple PRs then I can do so, perhaps we can get guidance from other reviewers?
On top of the prior usage in the fork mentioned previously, I have been running the original init script myself on multiple devices (WiFi routers, RTL838x switches and x86_64) through 22.03, 23.05 RC releases, 23.05.0 and 23.05.2 with no issues.
The huge amount of resistance is perplexing given that the lldpd init script is currently broken, ignoring all major options and binding to all interfaces regardless of what is set in UCI. We are not trying to sabotage, we are trying to fix a substantial gap.

stintel commented:
Both https://openwrt.org/submitting-patches#creating_a_patch and https://openwrt.org/submitting-patches#separate_your_changes tell you to split changes into multiple patches.

Commit:
lldpd: Init config read on reload

Init script reload with trigger to detect config file update.
Reload command added to attempt non-impactful lldpd reload where
lldpcli can be used to update config without process restart.
Config hash function used to track whether process restart is needed.

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: LLDPD binds to only specified interfaces

Bind to the configured system interfaces only. Switchport interfaces
are no longer ignored and uci interface values for LLDPD are honored.

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: Allow neighbour filtering

add filter option to init script.

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: Init adds no-version option

add option to disable advertising kernel version

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: option to disable LLDP-MED inventory TLV

add option to disable LLDP-MED inventory TLV transmission

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: add LLDP MED options

add option to enable LLDP MED fast-start and set fast-start timer

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: add agent-type option

add option to set agent-type to control propogation

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: add portidsubtype option

add option portidsubtype to correct port identifiers and descriptions

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: allow disabling LLDP protcol

add option to allow LLDP disabling while using other supported protocols

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: set CDP version and allow forcing CDP on

add option to specify CDPv1 or CDPv2 and separately enable or force each

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: add option to force FDP on

add option to force FDP when no peers detected

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: add option to force SONMP enabled

add option to force SONMP to be enabled even when no peer detected

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: add option to set system platform

add option to override system platform instead of using kernel name

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: add option for tx delay and tx hold

add option to set LLDP transmit delay, hold timers to set update frequency

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: Update Makefile package release

increment Makefile package release to reflect changes to init script

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

howels commented:
> Both https://openwrt.org/submitting-patches#creating_a_patch and https://openwrt.org/submitting-patches#separate_your_changes tell you to split changes into multiple patches.
> 
> You list 6 different, unrelated changes in the commit message. So that should be 6 commits. This makes review super easy. With the single diff and 6 different changes in the subject, the reviewer has to figure out which change belongs to which line in the commit message.
> 
> There are good reasons our commit guidelines, based on the commit guidelines of the Linux kernel, are the way they are. What perplexes me is that people keep doing things differently, and when I raise concerns and ask to do things properly, I get a thumbs down.
> 
> As for battle tested in Tano, I have no idea what Tano is, never heard of it, so that argument has no value to me. Then it also raises the question: why didn't the original author submit their changes to us?
> 
> And finally, I use lldpd on a plethora of OpenWrt devices, and claiming it is completely broken is just false.

I have gone ahead and split this into a commit for each distinct change incorporated here. I'm glad that you have better luck with LLDPD but I have had to resort to writing configuration directly to /etc/lldpd/ as the UCI values for interface binding were being ignored with the current init script.
Sorry for getting annoyed, I was frustrated by the slow progress but understand the burden on maintainers. While splitting I have rewritten these changes to be as obvious as possible in relation to the existing code. I hope that this makes future maintenance easier.

howels commented:
added commit descriptions in addition to the message, workflow seems to be failing without the second line.

stintel commented:
Thanks for splitting up. Much easier for me to review. Unfortunately you know how 16 commits that all have the same commit subject. This is not really acceptable. It was actually much better like this:

```
[@howels](https://github.com/howels) pushed 16 commits.

    https://github.com/openwrt/openwrt/commit/bed6cbac14af13c1ec45bd85fe10007fdc418869 lldpd: Init script reload on config file update
    https://github.com/openwrt/openwrt/commit/d05955d86f4b47f6047e6692e69b556cb309e349 lldpd: Bind to the configured system interfaces only
    https://github.com/openwrt/openwrt/commit/ad1bb5e2f221191c4826f18114a608f1bedeb8ac lldpd: add filter option
    https://github.com/openwrt/openwrt/commit/45a004144b84d8ea1a3d16f2885cf125b8e2c5e3 lldpd: add no-version option
    https://github.com/openwrt/openwrt/commit/87abd3b3d1e9781109bdb1ae5aaad3c814353ecf lldpd: add option to disable LLDP-MED inventory transmission
    https://github.com/openwrt/openwrt/commit/e4caa1ba97c045373664b875f8a10382923d82fc lldpd: add option to enable LLDP MED fast-start and set fast-start timer
    https://github.com/openwrt/openwrt/commit/7d8021f4d595b0a9b1a05c7dbb509fc7e9886280 lldpd: add option to set agent-type to control propogation
    https://github.com/openwrt/openwrt/commit/57a77e021eee38da76edb0a6aebba854c691ea0c lldpd: add option to set portidsubtype to correct port identifiers and descriptions
    https://github.com/openwrt/openwrt/commit/d52f7323c4f95dc6dd47789fa979e6f32e705720 lldpd: add option to allow LLDP disabling (when using CDP or other protocols) or to force LLDP
    https://github.com/openwrt/openwrt/commit/680c4192aa4cebc3ddb8580b396e77759e597f0c lldpd: add option to specify CDPv1 or CDPv2 and separately enable or force each
    https://github.com/openwrt/openwrt/commit/2ce8dfdd16069e082cb2e37fbfc1dd97f7f0813a lldpd: add option to force FDP
    https://github.com/openwrt/openwrt/commit/0419bd079b059aedbf7c6ece8a260bc386ef322d lldpd: add option to force SONMP
    https://github.com/openwrt/openwrt/commit/f7280d84d6306b623100c44112a0838222b2ca66 lldpd: add option to force EDP
    https://github.com/openwrt/openwrt/commit/ffb09542efcd9fa3529d4b147f06b05473c1e127 lldpd: add option to set system platform
    https://github.com/openwrt/openwrt/commit/a13a05ee30e140c61bcf362bcf3e3e2ec70cccee lldpd: add option to set LLDP transmit delay and hold value
    https://github.com/openwrt/openwrt/commit/dd158fd4e5f2f407b546dc14efe9865a252a705a lldpd: increment Makefile package release
```

If the workflow is failing because there is only a subject, we need to fix that workflow. If you can fully explain what the commit does in the subject, I do not see the need to have additional. Here's a few examples of commits that have done this recently:
3ed7abfc5af63c7fe1137a0bb78d95eb6f2455ef
00256d86dc2041882509fd3c0261905d326b4ff0
9ab259ab3435924726f6a603fefe67fd5026e936
2723f16ddadc6400cdbbadd0a54bc97b309597e1
684d1a5c359f6bf022b7f526672373d69967bced

This is perfectly acceptable, and it is done regularly, so the CI should not fail on such commits. It is up to the reviewer to decide if the subject says enough about the commit.

Why having the same commit subject is not OK... Just look at https://github.com/howels/openwrt/commits/lldpd-refactor. You have no idea what is actually done in any of those commits, without clicking open the rest of the message. The commit subject must not be generalized.

howels commented:
> Thanks for splitting up. Much easier for me to review. Unfortunately you know how 16 commits that all have the same commit subject. This is not really acceptable. It was actually much better like this:
> 
> ```
> [@howels](https://github.com/howels) pushed 16 commits.
> 
>     https://github.com/openwrt/openwrt/commit/bed6cbac14af13c1ec45bd85fe10007fdc418869 lldpd: Init script reload on config file update
>     https://github.com/openwrt/openwrt/commit/d05955d86f4b47f6047e6692e69b556cb309e349 lldpd: Bind to the configured system interfaces only
>     https://github.com/openwrt/openwrt/commit/ad1bb5e2f221191c4826f18114a608f1bedeb8ac lldpd: add filter option
>     https://github.com/openwrt/openwrt/commit/45a004144b84d8ea1a3d16f2885cf125b8e2c5e3 lldpd: add no-version option
>     https://github.com/openwrt/openwrt/commit/87abd3b3d1e9781109bdb1ae5aaad3c814353ecf lldpd: add option to disable LLDP-MED inventory transmission
>     https://github.com/openwrt/openwrt/commit/e4caa1ba97c045373664b875f8a10382923d82fc lldpd: add option to enable LLDP MED fast-start and set fast-start timer
>     https://github.com/openwrt/openwrt/commit/7d8021f4d595b0a9b1a05c7dbb509fc7e9886280 lldpd: add option to set agent-type to control propogation
>     https://github.com/openwrt/openwrt/commit/57a77e021eee38da76edb0a6aebba854c691ea0c lldpd: add option to set portidsubtype to correct port identifiers and descriptions
>     https://github.com/openwrt/openwrt/commit/d52f7323c4f95dc6dd47789fa979e6f32e705720 lldpd: add option to allow LLDP disabling (when using CDP or other protocols) or to force LLDP
>     https://github.com/openwrt/openwrt/commit/680c4192aa4cebc3ddb8580b396e77759e597f0c lldpd: add option to specify CDPv1 or CDPv2 and separately enable or force each
>     https://github.com/openwrt/openwrt/commit/2ce8dfdd16069e082cb2e37fbfc1dd97f7f0813a lldpd: add option to force FDP
>     https://github.com/openwrt/openwrt/commit/0419bd079b059aedbf7c6ece8a260bc386ef322d lldpd: add option to force SONMP
>     https://github.com/openwrt/openwrt/commit/f7280d84d6306b623100c44112a0838222b2ca66 lldpd: add option to force EDP
>     https://github.com/openwrt/openwrt/commit/ffb09542efcd9fa3529d4b147f06b05473c1e127 lldpd: add option to set system platform
>     https://github.com/openwrt/openwrt/commit/a13a05ee30e140c61bcf362bcf3e3e2ec70cccee lldpd: add option to set LLDP transmit delay and hold value
>     https://github.com/openwrt/openwrt/commit/dd158fd4e5f2f407b546dc14efe9865a252a705a lldpd: increment Makefile package release
> ```
> 
> If the workflow is failing because there is only a subject, we need to fix that workflow. If you can fully explain what the commit does in the subject, I do not see the need to have additional. Here's a few examples of commits that have done this recently: [3ed7abf](https://github.com/openwrt/openwrt/commit/3ed7abfc5af63c7fe1137a0bb78d95eb6f2455ef) [00256d8](https://github.com/openwrt/openwrt/commit/00256d86dc2041882509fd3c0261905d326b4ff0) [9ab259a](https://github.com/openwrt/openwrt/commit/9ab259ab3435924726f6a603fefe67fd5026e936) [2723f16](https://github.com/openwrt/openwrt/commit/2723f16ddadc6400cdbbadd0a54bc97b309597e1) [684d1a5](https://github.com/openwrt/openwrt/commit/684d1a5c359f6bf022b7f526672373d69967bced)
> 
> This is perfectly acceptable, and it is done regularly, so the CI should not fail on such commits. It is up to the reviewer to decide if the subject says enough about the commit.
> 
> Why having the same commit subject is not OK... Just look at https://github.com/howels/openwrt/commits/lldpd-refactor. You have no idea what is actually done in any of those commits, without clicking open the rest of the message. The commit subject must not be generalized.
> 
> Then finally, the subjects should be limited in length. See https://openwrt.org/submitting-patches#submission_guidelines "commit subject".

Rewrote the subject and messages again - pretty sure I am within message and subject lengths here and hope that the message and subject inform the reader more easily as to the content of the changes.

stintel commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -90,13 +90,27 @@ write_lldpd_conf()
 	[ -n "$lldp_hostname" ] && echo "configure system hostname" "\"$lldp_hostname\"" >> "$LLDPD_CONF"
 	[ -n "$lldp_mgmt_ip" ] && echo "configure system ip management pattern" "\"$lldp_mgmt_ip\"" >> "$LLDPD_CONF"
 	[ -n "$lldp_syscapabilities" ] && echo "configure system capabilities enabled" "\"$lldp_syscapabilities\"" >> "$LLDPD_CONF"
+		if [ $lldpmed_fast_start -gt 0 ]; then
```
Indentation seems wrong here.

howels replied:
thanks, corrected now.

stintel reviewed (changes requested):
I've reviewed all but the first commit in this PR. There's one minor issue with indentation. As for the first commit, this could use a bit of detailed explanation of what this does exactly and why it is needed.

howels commented:
> I've reviewed all but the first commit in this PR. There's one minor issue with indentation. As for the first commit, this could use a bit of detailed explanation of what this does exactly and why it is needed.

Added a more thorough explanation. The new reload behaviour is intended to detect config file changes and avoid a lldpd stop/start where those changes can be made via lldpcli.

stintel commented:
Merged in my staging tree, will test locally before pushing to master.

stintel commented:
```Tue Dec 19 03:09:35 2023 daemon.notice procd: /etc/rc.d/S90lldpd: /etc/rc.common: /etc/rc.d/S90lldpd: line 179: syntax error: unexpected "else" (expecting "}")```

stintel commented:
Apparently something went wrong during my build, the init script that ended up on my device had a different checksum than the one in the git repo. Now the next problem:

```
Tue Dec 19 04:34:57 2023 daemon.info procd: Instance lldpd::instance1 s in a crash loop 6 crashes, 0 seconds since last crash
```

After enabling stdout and stderr in procd I see this:

```
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: /usr/sbin/lldpd: unrecognized option: c
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: Usage:   lldpd [OPTIONS ...]
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: Version: lldpd 1.0.17
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -d       Do not daemonize.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -r       Receive-only mode
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -i       Disable LLDP-MED inventory TLV transmission.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -k       Disable advertising of kernel release, version, machine.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -S descr Override the default system description.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -P name  Override the default hardware platform.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -m IP    Specify the IP management addresses of this system.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -u file  Specify the Unix-domain socket used for communication with lldpctl(8).
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -H mode  Specify the behaviour when detecting multiple neighbors.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -I iface Limit interfaces to use.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -C iface Limit interfaces to use for computing chassis ID.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -L path  Override path for lldpcli command.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -O file  Override default configuration locations processed by lldpcli(8) at start.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -M class Enable emission of LLDP-MED frame. 'class' should be one of:
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:              1 Generic Endpoint (Class I)
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:              2 Media Endpoint (Class II)
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:              3 Communication Device Endpoints (Class III)
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:              4 Network Connectivity Device
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -x       Enable SNMP subagent.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -X sock  Specify the SNMP subagent socket.
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:
Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: see manual page lldpd(8) for more information
```

stintel commented:
The default config has

```
        option enable_cdp 1
        option enable_fdp 1
        option enable_sonmp 1
        option enable_edp 1
```

That won't work when building lldpd with support for those protocols enabled, that was already a problem before these changes I think. After fixing that problem, lldpd is running, but I'm not seeing it on the neighbor. I'll have to do some more testing later.

systemcrash commented:
> The default config has
> 
> ```
>         option enable_cdp 1
>         option enable_fdp 1
>         option enable_sonmp 1
>         option enable_edp 1
> ```
> 
> That won't work when building lldpd with support for those protocols enabled, but that was already a problem before these changes I think. After fixing that problem, lldpd is running, but I'm not seeing it on the neighbor. I'll have to do some more testing later.

I recall @npeca75 had a problem with [not seeing neighbours](https://forum.openwrt.org/t/lldp-or-dsa-bug-on-mt7621/175442/8).

howels commented:
> Apparently something went wrong during my build, the init script that ended up on my device had a different checksum than the one in the git repo. Now the next problem:
> 
> ```
> Tue Dec 19 04:34:57 2023 daemon.info procd: Instance lldpd::instance1 s in a crash loop 6 crashes, 0 seconds since last crash
> ```
> 
> After enabling stdout and stderr in procd I see this:
> 
> ```
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: /usr/sbin/lldpd: unrecognized option: c
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: Usage:   lldpd [OPTIONS ...]
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: Version: lldpd 1.0.17
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -d       Do not daemonize.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -r       Receive-only mode
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -i       Disable LLDP-MED inventory TLV transmission.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -k       Disable advertising of kernel release, version, machine.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -S descr Override the default system description.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -P name  Override the default hardware platform.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -m IP    Specify the IP management addresses of this system.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -u file  Specify the Unix-domain socket used for communication with lldpctl(8).
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -H mode  Specify the behaviour when detecting multiple neighbors.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -I iface Limit interfaces to use.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -C iface Limit interfaces to use for computing chassis ID.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -L path  Override path for lldpcli command.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -O file  Override default configuration locations processed by lldpcli(8) at start.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -M class Enable emission of LLDP-MED frame. 'class' should be one of:
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:              1 Generic Endpoint (Class I)
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:              2 Media Endpoint (Class II)
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:              3 Communication Device Endpoints (Class III)
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:              4 Network Connectivity Device
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -x       Enable SNMP subagent.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: -X sock  Specify the SNMP subagent socket.
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]:
> Tue Dec 19 04:36:49 2023 daemon.err lldpd[2595]: see manual page lldpd(8) for more information
> ```

Are you running an older release? On 23.05.2 I see this:
```root@edgerouter-x-3:~# lldpd -h
Usage:   lldpd [OPTIONS ...]
Version: lldpd 1.0.17

-d       Do not daemonize.
-r       Receive-only mode
-i       Disable LLDP-MED inventory TLV transmission.
-k       Disable advertising of kernel release, version, machine.
-S descr Override the default system description.
-P name  Override the default hardware platform.
-m IP    Specify the IP management addresses of this system.
-u file  Specify the Unix-domain socket used for communication with lldpctl(8).
-H mode  Specify the behaviour when detecting multiple neighbors.
-I iface Limit interfaces to use.
-C iface Limit interfaces to use for computing chassis ID.
-L path  Override path for lldpcli command.
-O file  Override default configuration locations processed by lldpcli(8) at start.
-M class Enable emission of LLDP-MED frame. 'class' should be one of:
             1 Generic Endpoint (Class I)
             2 Media Endpoint (Class II)
             3 Communication Device Endpoints (Class III)
             4 Network Connectivity Device

Additional protocol support.
-c       Enable the support of CDP protocol. (Cisco)
-e       Enable the support of EDP protocol. (Extreme)
-f       Enable the support of FDP protocol. (Foundry)
-s       Enable the support of SONMP protocol. (Nortel)
```
And starting lldpd with `-c` works correctly. This is the mips kernel on Ubiquiti EdgeRouter.

stintel commented:
> Have you built your own LLDPD binary?

Yes, with support for those proprietary protocols disabled. That explains the -c not working.

The bigger problem is that the init script ends up broken when some of the default build options are disabled.
Please have a look at https://github.com/openwrt/openwrt/blob/main/package/network/services/lldpd/Makefile#L70-L84 - the sed commands executed there break the init script.

I build with these options in .config:
```
CONFIG_LLDPD_WITH_PRIVSEP=y
# CONFIG_LLDPD_WITH_CDP is not set
# CONFIG_LLDPD_WITH_FDP is not set
# CONFIG_LLDPD_WITH_EDP is not set
# CONFIG_LLDPD_WITH_SONMP is not set
CONFIG_LLDPD_WITH_LLDPMED=y
CONFIG_LLDPD_WITH_DOT1=y
CONFIG_LLDPD_WITH_DOT3=y
CONFIG_LLDPD_WITH_CUSTOM=y
# CONFIG_LLDPD_WITH_JSON is not set
CONFIG_LLDPD_WITH_SNMP=y
```

The resulting init script errors:
```
/etc/rc.common: /etc/init.d/lldpd: line 179: syntax error: unexpected "else" (expecting "}")
```

howels commented:
> > Have you built your own LLDPD binary?
> 
> Yes, with support for those proprietary protocols disabled. That explains the -c not working.
> 
> The bigger problem is that the init script ends up broken when some of the default build options are disabled. Please have a look at https://github.com/openwrt/openwrt/blob/main/package/network/services/lldpd/Makefile#L70-L84 - the sed commands executed there break the init script.
> 
> I build with these options in .config:
> 
> ```
> CONFIG_LLDPD_WITH_PRIVSEP=y
> # CONFIG_LLDPD_WITH_CDP is not set
> # CONFIG_LLDPD_WITH_FDP is not set
> # CONFIG_LLDPD_WITH_EDP is not set
> # CONFIG_LLDPD_WITH_SONMP is not set
> CONFIG_LLDPD_WITH_LLDPMED=y
> CONFIG_LLDPD_WITH_DOT1=y
> CONFIG_LLDPD_WITH_DOT3=y
> CONFIG_LLDPD_WITH_CUSTOM=y
> # CONFIG_LLDPD_WITH_JSON is not set
> CONFIG_LLDPD_WITH_SNMP=y
> ```
> 
> The resulting init script errors:
> 
> ```
> /etc/rc.common: /etc/init.d/lldpd: line 179: syntax error: unexpected "else" (expecting "}")
> ```

Looked for a clean solution and think I found one. Build-time options are now set as variables in the init script and we only use init functionality when that `CONFIG_LLDPD_WITH_` argument was set at build time. This is slightly cleaner than relying on `sed` to remove all instances of a particular term like "cdp" throughout the init script.
I have relocated some init script commands to group variables relating to a specific feature like cdp into one block and make the if statements less numerous.

systemcrash commented:
Effective and elegant.

stintel commented:
> Looked for a clean solution and think I found one. Build-time options are now set as variables in the init script and we only use init functionality when that `CONFIG_LLDPD_WITH_` argument was set at build time. This is slightly cleaner than relying on `sed` to remove all instances of a particular term like "cdp" throughout the init script.
> I have relocated some init script commands to group variables relating to a specific feature like cdp into one block and make the if statements less numerous.

I was thinking of a similar, yet slightly different approach: include files in the package to indicate if something is supported. E.g. /usr/share/lldpd/features/cdp. This way, we don't have to modify the init script during build at all. This feels slightly less hacky. Curious to hear what people think about this.

howels commented:
> > Looked for a clean solution and think I found one. Build-time options are now set as variables in the init script and we only use init functionality when that `CONFIG_LLDPD_WITH_` argument was set at build time. This is slightly cleaner than relying on `sed` to remove all instances of a particular term like "cdp" throughout the init script.
> > I have relocated some init script commands to group variables relating to a specific feature like cdp into one block and make the if statements less numerous.
> 
> I was thinking of a similar, yet slightly different approach: include files in the package to indicate if something is supported. E.g. /usr/share/lldpd/features/cdp. This way, we don't have to modify the init script during build at all. This feels slightly less hacky. Curious to hear what people think about this.

I could do either way. I was trying to make the build options as transparent as possible so a glance at the init script and you know why they are set or not set. But if a feature filesystem is a better strategic fit with what other packages have done I am happy to follow suit.

systemcrash commented:
> I could do either way. I was trying to make the build options as transparent as possible so a glance at the init script and you know why they are set or not set. But if a feature filesystem is a better strategic fit with what other packages have done I am happy to follow suit.

Is querying the actual binary to see what it supports ( what's compiled in ) an option? Stash those lines as variables.

howels commented:
> > I could do either way. I was trying to make the build options as transparent as possible so a glance at the init script and you know why they are set or not set. But if a feature filesystem is a better strategic fit with what other packages have done I am happy to follow suit.
> 
> Is querying the actual binary to see what it supports ( what's compiled in ) an option? Stash those lines as variables.

Might as well use the build variables when they are available, trying to recreate the build variables by scraping the lldpd binary's usage output is getting the same data by a less reliable route.

systemcrash commented:
Are build variables available when installed on a device? I know the
supported protocos are packed reliably in the installed binary.

howels commented:
> Are build variables available when installed on a device? I know the supported protocos are packed reliably in the installed binary.

Check the latest commit here - I simply placed the lldpd build options as variables in the init script to enable or disable the various feature blocks.

systemcrash commented:
Oh OK. I thought something else had changed since then. Variables in the init script is OK. The only time it is modified is during install right?

howels commented:
> Oh OK. I thought something else had changed since then. Variables in the init script is OK. The only time it is modified is during install right?

Init script is only modified during the package building phase. After that it's static.

howels commented:
Now that we are moving past the winter holidays, @systemcrash and @stintel please let me know the next steps you would like here.

systemcrash commented:
I'm OK with merging if all concerns have been addressed. Also okay with
looping in some more reviewers.

Perhaps @npeca75 might have a few words in lldpd.

Npeca75 commented:
> Perhaps @Npeca75 might have a few words in lldpd.

hi
yes, i see that init script is modified for more config options
nice job

but my main problem was with separate GMAC1 and MT7621 so i reverted DTS to GMAC0 only and it is working perfectly fine for me as i using OWRT routers as DumbAPs, therefore no need for WAN/NAT/Routing speed up with GMAC1

anyway ... nice to have more config options
p.s. i don't see option for custom-tlv
for ex: `configure ports eth0.1 lldp custom-tlv add oui 00,80,c2 subtype 1 oui-info '00,01'`
maybe i am wrong ?

systemcrash commented:
> for ex: `configure ports eth0.1 lldp custom-tlv add oui 00,80,c2 subtype 1 oui-info '00,01'`
> maybe i am wrong ?

Not wrong - perhaps those can find a place in this PR, given how long it takes to get changes in. @howels

howels commented:
> > Perhaps @Npeca75 might have a few words in lldpd.
> 
> hi yes, i see that init script is modified for more config options nice job
> 
> but my main problem was with separate GMAC1 and MT7621 so i reverted DTS to GMAC0 only and it is working perfectly fine for me as i using OWRT routers as DumbAPs, therefore no need for WAN/NAT/Routing speed up with GMAC1
> 
> anyway ... nice to have more config options p.s. i don't see option for custom-tlv for ex: `configure ports eth0.1 lldp custom-tlv add oui 00,80,c2 subtype 1 oui-info '00,01'` maybe i am wrong ?

I'm having a look and I'll see how it can be done.

howels commented:
> > for ex: `configure ports eth0.1 lldp custom-tlv add oui 00,80,c2 subtype 1 oui-info '00,01'`
> > maybe i am wrong ?
> 
> Not wrong - perhaps those can find a place in this PR, given how long it takes to get changes in. @howels

Could we try and get this merged so that the Luci app can be enabled and we can bind to the selected interfaces please? Then I can focus on the additional per-port features which look like a sizeable addition for a later PR.

systemcrash commented:
Sure. I think that's the most sensible way forward.



> Could we try and get this merged so that the Luci app can be enabled and
> we can bind to the selected interfaces please? Then I can focus on the
> additional per-port features which look like a sizeable addition for a
> later PR.
>
>
>

systemcrash commented:
@stintel ?

stintel commented:
Apologies for not responding sooner, I've been in between travelling to visit family for the holidays, starting a new project in Rust which was a big context switch, friends visiting and more travelling to ski resorts.

So the problem with fixing things in the last commit is that as of some commit in this PR, the init script is potentially broken, and only fixed at the end of the series. Ideally this should be avoided.

howels commented:
> Apologies for not responding sooner, I've been in between travelling to visit family for the holidays, starting a new project in Rust which was a big context switch, friends visiting and more travelling to ski resorts.
> 
> So the problem with fixing things in the last commit is that as of some commit in this PR, the init script is potentially broken, and only fixed at the end of the series. Ideally this should be avoided.

Thanks for responding. I can make "lldpd: init only uses options enabled in build" an earlier commit in the PR?

stintel commented:
> Thanks for responding. I can make "lldpd: init only uses options enabled in build" an earlier commit in the PR?
> If anything this is easier just squashed into a single commit.

That would work. Squashing things would make this a single large commit again, so all the work of splitting things would have been for nothing, not good for bisecting, etc. I wish github did not offer this damn option.

Commit:
lldpd: only use snmp options when compiled in

prevent SNMP options being passed unless lldpd supports them

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

Commit:
lldpd: add option to force EDP

allow EDP support if compiled and add force EDP option

Signed-off-by: Stephen Howell <howels@allthatwemight.be>

howels commented:
> > Thanks for responding. I can make "lldpd: init only uses options enabled in build" an earlier commit in the PR?
> > If anything this is easier just squashed into a single commit.
> 
> That would work. Squashing things would make this a single large commit again, so all the work of splitting things would have been for nothing, not good for bisecting, etc. I wish github did not offer this damn option.

I have gone and updated each commit so that we are now looking for build flags in the same commit that deals with that particular argument. This should allow commits to be applied in whatever order you choose.

stintel commented:
Just returned from another snowboarding weekend, I'll try test this tomorrow and get it merged.

stintel commented:
Pushed to master.

howels commented:
> Pushed to master.

Thanks for your help and continued patience.

stintel commented:
> Thanks for your help and continued patience.

Same to you ;-)

systemcrash commented on package/network/services/lldpd/files/lldpd.init:
```diff
@@ -48,75 +115,239 @@ write_lldpd_conf()
-	[ $enable_sonmp -gt 0 ] && procd_append_param command '-s'
-	[ $enable_edp -gt 0 ] && procd_append_param command '-e'
 	[ $readonly_mode -gt 0 ] && procd_append_param command '-r'
+	[ $lldp_no_version -gt 0 ] && procd_append_param commanpackage/network/services/lldpd/Makefile package/network/services/lldpd/files/lldpd.initd '-k'
```
Late breaking, but should that be as so?:
```
	[ $lldp_no_version -gt 0 ] && procd_append_param commonpackage/network/services/lldpd/Makefile package/network/services/lldpd/files/lldpd.initd '-k'
```

In fact, is this line right at all or did this slip in from some fumble?

systemcrash replied:
Seems like it should be
```sh
	[ $lldp_no_version -gt 0 ] && procd_append_param command '-k'
```

systemcrash replied:
Yep - 100% sure with testing now. This row is a fumble and needs patching.

systemcrash commented:
See #14583

systemcrash commented:
@Npeca75 
> anyway ... nice to have more config options p.s. i don't see option for custom-tlv for ex: `configure ports eth0.1 lldp custom-tlv add oui 00,80,c2 subtype 1 oui-info '00,01'` maybe i am wrong ?

I've implemented the basics in the init script in #14872.

The GUI is in [a stack of updates I have queued for luci](https://github.com/openwrt/luci/compare/master...systemcrash:luci:marek22k/master) - they depend on #14867. 

If you like what you see, please go and give your approval on those PRs ( -> Files Changed -> Review -> Approve ). Great if you can test that code also. Another data-point that I've tested enough :)
